File: src/workers/claude.ts
import { createInterface } from "node:readline";
import type { TurnRequest, TurnResult, WorkerHarness } from "./harness.js";

// Claude Code worker harness over the live-verified control surface
// (PLAN § Worker Harness):
//   start:  claude --print --output-format stream-json --verbose <perm> <prompt>
//   resume: same, plus --resume <session_id>
// Non-interactive runs MUST set a permission flag or they hang on permission
// prompts. In Docker the container plus egress proxy are the boundary, so
// Claude's own permission prompts are disabled; on the host, acceptEdits
// allows file edits but nothing broader.
//
// Claude emits no dedicated file-change event; changed files are derived
// from tool_use inputs (Write/Edit/MultiEdit/NotebookEdit), with the
// workspace git fallback covering anything shell commands touched.

interface ClaudeLine {
  [key: string]: unknown;
}

const EDIT_TOOLS = new Set(["Write", "Edit", "MultiEdit", "NotebookEdit"]);

function extractEditedFiles(message: ClaudeLine, workspacePath: string, into: Set<string>): void {
  const content = message["content"];
  if (!Array.isArray(content)) return;
  for (const block of content) {
    const b = block as ClaudeLine;
    if (b["type"] !== "tool_use" || !EDIT_TOOLS.has(String(b["name"]))) continue;
    const input = (b["input"] ?? {}) as ClaudeLine;
    const path = input["file_path"] ?? input["notebook_path"];
    if (typeof path !== "string") continue;
    const prefix = workspacePath.endsWith("/") ? workspacePath : `${workspacePath}/`;
    into.add(path.startsWith(prefix) ? path.slice(prefix.length) : path);
  }
}

export interface ClaudeHarnessOptions {
  /** Model to request, passed as `--model`. */
  model?: string;
}

export class ClaudeHarness implements WorkerHarness {
  readonly name = "claude";

  constructor(private readonly options: ClaudeHarnessOptions = {}) {}

  async runTurn(request: TurnRequest): Promise<TurnResult> {
    const args = ["claude", "--print", "--output-format", "stream-json", "--verbose"];
    if (this.options.model) args.push("--model", this.options.model);
    if (request.runner.kind === "docker") {
      args.push("--dangerously-skip-permissions");
    } else {
      args.push("--permission-mode", "acceptEdits");
    }
    if (request.nativeSessionId) args.push("--resume", request.nativeSessionId);
    args.push(request.prompt);

    const worker = await request.runner.start({ argv: args });

    return new Promise<TurnResult>((resolve, reject) => {
      let sessionId: string | undefined = request.nativeSessionId;
      let response = "";
      let usage: unknown;
      let resultError: string | undefined;
      const changedFiles = new Set<string>();
      let stderrTail = "";
      let settled = false;

      const onAbort = () => worker.kill();
      request.signal.addEventListener("abort", onAbort, { once: true });

      const rl = createInterface({ input: worker.stdout });
      rl.on("line", (line) => {
        if (line.trim() === "") return;
        let obj: ClaudeLine;
        try {
          obj = JSON.parse(line) as ClaudeLine;
        } catch {
          request.onEvent({ kind: "unparsed_output", payload: { line } });
          return;
        }
        // Event kinds quote Claude's own line types verbatim
        // (system, assistant, user, result, rate_limit_event, ...).
        const kind = typeof obj["type"] === "string" ? (obj["type"] as string) : "unknown";
        request.onEvent({ kind, payload: obj });

        if (typeof obj["session_id"] === "string") sessionId = obj["session_id"] as string;
        if (kind === "assistant" && obj["message"]) {
          extractEditedFiles(
            obj["message"] as ClaudeLine,
            request.runner.workspacePath,
            changedFiles,
          );
        }
        if (kind === "result") {
          if (typeof obj["result"] === "string") response = obj["result"] as string;
          if (obj["usage"] !== undefined) usage = obj["usage"];
          if (obj["is_error"] === true) {
            resultError = typeof obj["result"] === "string"
              ? (obj["result"] as string)
              : `claude reported ${String(obj["subtype"] ?? "an error")}`;
          }
        }
      });

      worker.stderr.on("data", (chunk: Buffer) => {
        stderrTail = (stderrTail + chunk.toString("utf8")).slice(-4096);
      });

      const settle = (fn: () => void) => {
        if (settled) return;
        settled = true;
        request.signal.removeEventListener("abort", onAbort);
        fn();
      };

      worker.exited
        .then((code) => {
          rl.close();
          if (request.signal.aborted) {
            settle(() => reject(new Error("claude worker terminated by abort")));
            return;
          }
          if (code !== 0 || resultError) {
            const detail = resultError ?? stderrTail.trim();
            settle(() =>
              reject(
                new Error(`claude exited with code ${code}${detail ? `: ${detail}` : ""}`),
              ),
            );
            return;
          }
          settle(() =>
            resolve({
              response,
              ...(sessionId ? { nativeSessionId: sessionId } : {}),
              changedFiles: [...changedFiles],
              ...(usage !== undefined ? { usage } : {}),
              exitCode: code,
            }),
          );
        })
        .catch((err: Error) =>
          settle(() => reject(new Error(`failed to start claude worker: ${err.message}`))),
        );
    });
  }
}

File: src/workers/claude.rs
//! Claude Code worker harness over the live-verified control surface:
//!   start:  claude --print --output-format stream-json --verbose <perm> <prompt>
//!   resume: same, plus --resume <session_id>
//! Non-interactive runs MUST set a permission flag or they hang on permission
//! prompts. In Docker the container plus egress proxy are the boundary, so
//! Claude's own permission prompts are disabled; on the host, acceptEdits
//! allows file edits but nothing broader.
//!
//! Claude emits no dedicated file-change event; changed files are derived
//! from tool_use inputs (Write/Edit/MultiEdit/NotebookEdit), with the
//! workspace git fallback covering anything shell commands touched.

use std::collections::BTreeSet;

use async_trait::async_trait;
use serde_json::Value;

use super::codex::relative_to_workspace;
use super::harness::{TurnRequest, TurnResult, WorkerEvent, WorkerHarness, drive};
use super::runner::{RunnerKind, WorkerSpawnSpec};
use crate::domain::errors::ToolError;

const EDIT_TOOLS: [&str; 4] = ["Write", "Edit", "MultiEdit", "NotebookEdit"];

fn extract_edited_files(message: &Value, workspace_path: &str, into: &mut BTreeSet<String>) {
    let Some(Value::Array(content)) = message.get("content") else { return };
    for block in content {
        let is_edit = block.get("type").and_then(Value::as_str) == Some("tool_use")
            && block
                .get("name")
                .and_then(Value::as_str)
                .is_some_and(|name| EDIT_TOOLS.contains(&name));
        if !is_edit {
            continue;
        }
        let input = block.get("input").filter(|v| !v.is_null());
        let path = input.and_then(|i| {
            i.get("file_path").filter(|v| !v.is_null()).or_else(|| i.get("notebook_path"))
        });
        if let Some(path) = path.and_then(Value::as_str) {
            into.insert(relative_to_workspace(path, workspace_path));
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct ClaudeHarnessOptions {
    /// Model to request, passed as `--model`.
    pub model: Option<String>,
}

pub struct ClaudeHarness {
    options: ClaudeHarnessOptions,
}

impl ClaudeHarness {
    pub fn new(options: ClaudeHarnessOptions) -> ClaudeHarness {
        ClaudeHarness { options }
    }

    fn spawn_spec(&self, request: &TurnRequest<'_>) -> WorkerSpawnSpec {
        let mut argv: Vec<String> =
            ["claude", "--print", "--output-format", "stream-json", "--verbose"]
                .iter()
                .map(|s| s.to_string())
                .collect();
        if let Some(model) = &self.options.model {
            argv.extend(["--model".to_string(), model.clone()]);
        }
        if request.runner.kind() == RunnerKind::Docker {
            argv.push("--dangerously-skip-permissions".into());
        } else {
            argv.extend(["--permission-mode".to_string(), "acceptEdits".to_string()]);
        }
        if let Some(session) = &request.native_session_id {
            argv.extend(["--resume".to_string(), session.clone()]);
        }
        argv.push(request.prompt.clone());
        WorkerSpawnSpec { argv, env: Default::default() }
    }
}

#[async_trait]
impl WorkerHarness for ClaudeHarness {
    fn name(&self) -> &str {
        "claude"
    }

    async fn run_turn(&self, request: TurnRequest<'_>) -> Result<TurnResult, ToolError> {
        let worker = request.runner.start(self.spawn_spec(&request)).await?;

        let mut session_id = request.native_session_id.clone();
        let mut response = String::new();
        let mut usage: Option<Value> = None;
        let mut result_error: Option<String> = None;
        let mut changed_files = BTreeSet::new();
        let exit = drive::run(worker, &request.cancel, |line| {
            let Ok(obj) = serde_json::from_str::<Value>(line) else {
                (request.on_event)(WorkerEvent {
                    kind: "unparsed_output".into(),
                    payload: serde_json::json!({ "line": line }),
                });
                return;
            };
            // Event kinds quote Claude's own line types verbatim
            // (system, assistant, user, result, rate_limit_event, ...).
            let kind = obj.get("type").and_then(Value::as_str).unwrap_or("unknown").to_string();
            (request.on_event)(WorkerEvent { kind: kind.clone(), payload: obj.clone() });

            if let Some(id) = obj.get("session_id").and_then(Value::as_str) {
                session_id = Some(id.to_string());
            }
            if kind == "assistant"
                && let Some(message) = obj.get("message").filter(|m| !m.is_null())
            {
                extract_edited_files(message, request.runner.workspace_path(), &mut changed_files);
            }
            if kind == "result" {
                if let Some(text) = obj.get("result").and_then(Value::as_str) {
                    response = text.to_string();
                }
                if let Some(u) = obj.get("usage") {
                    usage = Some(u.clone());
                }
                if obj.get("is_error") == Some(&Value::Bool(true)) {
                    result_error = Some(match obj.get("result").and_then(Value::as_str) {
                        Some(text) => text.to_string(),
                        None => format!(
                            "claude reported {}",
                            obj.get("subtype").map(json_string).unwrap_or("an error".into())
                        ),
                    });
                }
            }
        })
        .await;

        if exit.aborted || exit.code != Some(0) || result_error.is_some() {
            return Err(drive::failure("claude", &exit, result_error.as_deref()));
        }
        Ok(TurnResult {
            response,
            native_session_id: session_id.filter(|s| !s.is_empty()),
            changed_files: changed_files.into_iter().collect(),
            usage,
        })
    }
}

/// `String(v)`.
fn json_string(value: &Value) -> String {
    match value {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}
